### DISCUSSION TITLE
Critical Rendering Path 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by airman5573** [woowacourse-study/2022-woowahan-bansanghwe#60] [Optimization]
# Critical Rendering Path

## 간단 설명
브라우저는 html코드를 파싱해 DOM Tree를 생성하고 css코드를 파싱해 CSSOM Tree를 생성합니다.  
그리고 그 두 트리를 합성해서 render tree를 만듭니다.  
이후에 reflow과정을 거쳐 각 노드의 위치값을 계산하고,  
마지막으로 repaint과정에서 화면에 노드들을 픽셀화 합니다.

## 상세 설명
브라우저의 메인 쓰레드는 html코드를 파싱하는 과정에서 div같은 태그를 보고 javascript 객체를 생성하고  
그것을 트리구로 형태로 연결시켜 DOM tree를 생성합니다.  
파싱 도중에 rendering blocking과 parsing blocking 이 두가지 상황이 발생 할 수 있습니다.  

### rendering blocking
rendering blocking은 스타일 자원을 불러올때 발생합니다.  
메인 쓰레드가 파싱 도중에 외부 css파일을 불러오는 link 태그를 만나면 **브라우저에게 네트워크 요청을 맡기고 파싱을 계속 하면서** DOM Tree를 구성해 나갑니다.  
다만, DOM Tree를 다 만들었다 하더라도, 바로 다음 단계로 넘어가지 않고 css파일이 다 다운로드 되고 CSSOM tree가 만들어지기 까지 **기다립니다**.   
CSSOM Tree까지 다 만들어지고 나서 메인 쓰레드는 DOM Tree와 CSSOM Tree의 각 노드들을 합성해 Render Tree를 만듭니다.  

### parsing blocking
parsing blocking은 스크립트 자원을 불러오거나 실행할때 발생합니다.  
메인 쓰레드가 파싱 도중에 스크립트를 불러오는 script태그를 만나면 **파싱을 멈추고** 브라우저에게 네트워크 요청을 맡기고 기다립니다.  
파일이 다 불러와 지면 메인쓰레드는 **자바스크립트 엔진으로 넘어가서** 스크립트를 실행합니다. 스크립트 실행이 **다 끝나면 그때 남은 파싱과정을 계속** 해나갑니다.

이렇게 DOM Tree와 CSSOM Tree를 만들고 합성 과정을 거쳐 Render Tree(layout tree)를 만들면,  
노드들의 절대적인 위치를 계산하는 reflow과정을 거칩니다.  
그리고 render tree를 바탕으로 **어떤 순서로 그릴지 계산하는** repaint 단계를 거칩니다.  
<img width="561" alt="sc_ 67" src="https://user-images.githubusercontent.com/68623798/188258270-c3d15de7-3d01-4c39-b8e9-4f38ef4be91a.png">

이렇게 DOM을 만들고, 각 노드의 스타일과 절대 위치값을 계산하고,  
어떤 순서로 그려야 할지도 파악 했다면 이제 **이 정보들을 스크린에 픽셀로 그리는 resterizing** 을 거칩니다.  
<img width="690" alt="sc_ 69" src="https://user-images.githubusercontent.com/68623798/188258287-dd299e8f-bb74-4178-8b6c-8f138b552850.png">

다만 브라우저는 효율적인 렌더링을 위해서 한 페이지를 한번에 레스터라이징 하는것이 아니라,  
필요에 따라 **페이지를 여러 레이어로 나누고**, 각 레이어를 레스터라이징 해놓은 후에, 그 레이어들을 **합성**해 frame으로 만듭니다.  
그리고 그 frame을 **GPU를 사용해** 화면에 그립니다.    
이 과정을 compisiting이라고 부릅니다. 이 compositing과정은 **main thread가 아닌 compositor thread**에서 진행되기 때문에 **스타일 계산이나 javascript 실행을 기다릴 필요 없어서 성능에 더 좋습니다**.
<img width="1160" alt="sc_ 71" src="https://user-images.githubusercontent.com/68623798/188258309-69a9a74e-a4c2-474f-9008-91722e84afe1.png">

## 참고자료
- [Braces to Pixels](https://alistapart.com/article/braces-to-pixels/)
- [Why are some animations slow?](https://web.dev/animations-overview/)  
- [Inside look at modern web browser (part 3)](https://developer.chrome.com/blog/inside-browser-part3/)

**Comment by JUDONGHYEOK** [comment]
리플로우와 리페인트의 비용은 무엇이 더 많이 발생하고 그 이유는 무엇일지?

**Comment by JUDONGHYEOK** [comment]
왜 CriticalRenderingPath라고 불리는지

**Comment by greenblues1190** [comment]
will-change를 달면 어떤 동작을 하나요?

**Comment by JUDONGHYEOK** [comment]
reflow를 막기 위해서 translate를 쓰는이유

**Comment by greenblues1190** [comment]
> 필요에 따라 페이지를 여러 레이어로 나누고

레이어는 어떤 기준으로 나눠지나요?

**Comment by JUDONGHYEOK** [comment]
javascript script async/defer

**Comment by greenblues1190** [comment]
async와 defer는 어떤 모듈에 적용하는 것이 좋을까요?

**Comment by JUDONGHYEOK** [comment]
DomContentLoaded vs Load

**Comment by greenblues1190** [comment]
레스터라이제이션을 GPU가 담당하는 이유가 무엇인가요?

**Comment by byhhh2** [comment]
CPU가 하기 적합한 일 vs GPU가 하기 적합한 일

**Comment by greenblues1190** [comment]
1 pixel은 몇 byte일까요?

**Comment by greenblues1190** [comment]
자바스크립트는 싱글 스레드인데 브라우저는 어떻게 멀티 스레드인가요?
